### ISSUE TITLE
False Syntax Error: The `super` keyword may only be preceded by other `super`s

### ISSUE THREAD

**Issue by TonalidadeHidrica** [rust-analyzer/rust-analyzer#10770]
There's a bug in the syntax parser.

### How to reproduce?

Create an arbitrary library crate with `cargo new my_project --lib` then edit `src/lib.rs` as follows:
```rust
pub fn main() {
    println!("Hello, world!");
}

pub mod submodule {
    use self::super::main;

    pub fn foo() {
        main();
    }
}
```

### What's expected?

No diagnostics shown.  Indeed, `cargo check` (and `cargo clippy`) shows nothing.

### What happens?

The following error is detected on "super" in the 6th line:
```
[rust-analyzer syntax-error] [E] Syntax Error: The `super` keyword may only be preceded by other `super`s
```

### How to workaround?

Replace `self::super` with `super`.  Perhaps they're identical (I'm not quite sure).

**Comment by lnicola** [comment]
So `rustc` complains about `use crate::submodule::super::main;`:

    `super` in paths can only be used in start position

but actually accepts `self::super::main;`. I'm not sure what the rules are here.

Anyway, this validation can be easily removed, but I wonder if we should update the check instead of removing it.

**Comment by TonalidadeHidrica** [comment]
It seems that `super` is accepted in non-initial positions if it has a form of `self::super::super::super::` (arb. number of times).  I asked [Rust Internals Discourse](https://internals.rust-lang.org/t/use-self-super-is-accepted-is-this-intended/15616) if this is an intended behavior.

```rust
pub fn main() {
    println!("Hello, world!");
}

pub mod submodule {
    // use super::main;  // OK
    use self::super::main;  // Somehow accepted
    // use crate::submodule::super::main;  // Error

    pub fn foo() {
        main();
    }

    pub mod yet_another_submodule {
        // use super::super::main; // OK
        // use self::super::super::main;  // Somehow accepted
        // use super::self::super::main; // Error

        pub fn foo() {
            main();
        }
    }
}
```

**Comment by TonalidadeHidrica** [comment]
@lnicola

I got an answer with the reference to [the code which does path resolutions](https://github.com/rust-lang/rust/blob/master/compiler/rustc_resolve/src/lib.rs#L2204).  As far as I can read, the special path components `self`, `super` and `crate` are allowed as the first item, or

* the item is `super` and
* All the preceding components are either `super` or `self`.

Therefore, the only allowed form is `self::super::super:: ...` or `super::super::super:: ...`.   Since the latter is already considered correct by `rust-analyzer`, I think the only bug we should fix is `self::super::super:: ...`, then the `rust-analyzer`'s algorithm should be fully compliant to that of `rustc`.

**Comment by lnicola** [comment]
On the other hand:

>  So the restriction regarding their positions in paths is artificial, and I'd personally just lift it entirely some day.

Maybe we should push for `rustc` to become more flexibile instead of matching its behavior?

**Comment by TonalidadeHidrica** [comment]
You mean, allowing redundant paths, something like `use super::submodule::super`?

**Comment by flodiebold** [comment]
Yeah, we could just allow a superset of rustc here. Maybe we could add a "simplify path" assist that gets rid of those supers/selfs; that would at the same time serve as a fix for the rustc error.

**Comment by lnicola** [comment]
Yeah, I think that's what petrochenkov meant there?

**Comment by TonalidadeHidrica** [comment]
Well yes, IDEs should be more error-prone than compilers, so that seems very reasonable choice.

**Comment by TonalidadeHidrica** [comment]
Wow, thanks for such a fast implementation!
